File: checks/ruby/lib/section508/checks/audio_control.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Autoplay media longer than a moment needs an immediate pause.
    class AudioControl
      def self.call(root)
        return [] unless root&.respond_to?(:css)

        media = Dom.elements(root, "audio, video")
        return ["This section has no audio."] if media.empty?

        media.each do |element|
          next unless Dom.has_attr?(element, "autoplay")
          next if Dom.has_attr?(element, "controls")

          if root.at_css("button, [aria-label]")
            pause = Dom.elements(root, "button, [aria-label]").any? do |node|
              /pause|stop|mute/i.match?("#{node.text} #{node["aria-label"]}")
            end
            next if pause
          end
          return ["Audio starts on its own and has no pause control."]
        end
        []
      end
    end
  end
end

File: checks/python/src/section508/checks/audio_control.py
import re

from section508.dom import attr_or, has_attr, select

PAUSE = re.compile(r"pause|stop|mute", re.I)


def check(root):
    if not root:
        return []

    media = select(root, "audio, video")
    if not media:
        return ["This section has no audio."]

    for element in media:
        if not has_attr(element, "autoplay"):
            continue
        if has_attr(element, "controls"):
            continue
        if select(root, "button, [aria-label]"):
            pause = any(
                PAUSE.search(f"{node.get_text() or ''} {attr_or(node, 'aria-label')}")
                for node in select(root, "button, [aria-label]")
            )
            if pause:
                continue
        return ["Audio starts on its own and has no pause control."]
    return []
